Generalize Rotosolve to arbitray-frequency QNodes (#2081)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

<h3>New features since last release</h3>

* The `RotosolveOptimizer` has been generalized to arbitrary frequency spectra
  in the cost function.
  [(#2081)](https://github.com/PennyLaneAI/pennylane/pull/2081)

  Previously, the `RotsolveOptimizer` was available for cost functions with
  frequency spectra that only contained integers, and the maximal frequency
  (instead of the number of frequencies) determined the cost of the optimization.
  Now arbitrary frequencies are supported.

  Consider the QNode
  ```python
  dev = qml.device("default.qubit", wires=2)

  @qml.qnode(dev)
  def qnode(x, Y):

>       qml.RX(2.5*x, wires=0)
>       qml.CNOT(wires=[0, 1])
>       qml.RZ(0.3*Y[0], wires=0)
>       qml.CRY(1.1*Y[1], wires=[1, 0])
>       return qml.expval(qml.PauliX(0)@qml.PauliZ(1))

Reviewer: ```suggestion
      qml.RX(2.5 * x, wires=0)
      qml.CNOT(wires=[0, 1])
      qml.RZ(0.3 * Y[0], wires=0)
      qml.CRY(1.1 * Y[1], wires=[1, 0])
      return qml.expval(qml.PauliX(0) @ qml.PauliZ(1))
```

  x = np.array(0.8, requires_grad=True)
  Y = np.array([-0.2, 1.5], requires_grad=True)
  ```

  Its frequency spectra can be easily obtained via `qml.fourier.qnode_spectrum`:
  ```pycon
  >>> spectra = qml.fourier.qnode_spectrum(qnode)(x, Y)
  >>> spectra
  {'x': {(): [-2.5, 0.0, 2.5]},
   'Y': {(0,): [-0.3, 0.0, 0.3], (1,): [-1.1, -0.55, 0.0, 0.55, 1.1]}}
  ```

  We may then initialize the `RotosolveOptimizer` and minimize the QNode cost function
  by providing this information about the frequency spectra. We also compare the cost at
  each step to the initial cost.
  ```pycon
  >>> cost_init = qnode(x, Y)
  >>> opt = qml.RotosolveOptimizer()
  >>> for _ in range(2):

>   ...     x, Y = opt.step(qnode, x, Y, spectra=spectra)

Reviewer: I am partly wondering why the spectra is not passed an initialization. I suppose it makes more sense to do this:

```python
>>> opt = qml.RotosolveOptimizer(qnode, spectra=spectra)
>>> for _ in range(2):
...     x, Y = opt.step(x, Y)
```

but this is a longer term breaking change, and connects to #1564

  ...     print(f"New cost: {np.round(qnode(x, Y), 3)}; Initial cost: {np.round(cost_init, 3)}")
  New cost: 0.0; Initial cost: 0.706
  New cost: -1.0; Initial cost: 0.706

>   ```

Reviewer: omg, in only two steps?! 😍 I know it is a very simple cost function, but it still blows my mind!

Reviewer: out of curiosity, is there a rough heuristic in terms of what type of variational algorithms we think rotosolve will perform well at?

Author: Haha, nice research project you got there :P :D 
I actually could imagine it to not be the stupidest idea to apply Rotosolve to QAOA, but I don't have hard reasons for this.

  The optimization with `RotosolveOptimizer` is performed in substeps. The minimal cost
  of these substeps can be retrieved by setting `full_output=True`.
  ```pycon
  >>> x = np.array(0.8, requires_grad=True)
  >>> Y = np.array([-0.2, 1.5], requires_grad=True)
  >>> cost_init = qnode(x, Y)
  >>> opt = qml.RotosolveOptimizer()
  >>> for _ in range(2):
  ...     x, Y = opt.step(qnode, x, Y, spectra=spectra)
  ...     print(f"New cost: {np.round(qnode(x, Y), 3)} reached via substeps {np.round(history, 3)}")
  New cost: 0.0 reached via substeps [-0.  0.  0.]

>   New cost: -1.0 reached via substeps [-0.276 -0.276 -1.   ]

Reviewer: maybe I am missing it, but where is `full_output=True` passed, and where is `history` defined?

Author: Oh, something must have gone wrong when inserting the code example. Thanks!

  ```
  However, note that these intermediate minimal values are evaluations of the

>   _reconstructions_ that Rotosolve creates and uses internally for the optimization,

Reviewer: ```suggestion
  *reconstructions* that Rotosolve creates and uses internally for the optimization,
```

>   and not of the original objective function. For noisy cost functions, these may
>   be significantly different.

Reviewer: ```suggestion
  and not of the original objective function. For noisy cost functions, these intermediate
  evaluations may differ significantly from evaluations of the original cost function.
```

maybe to be a little more explicit?

<h3>New features since last release</h3>

* The `RotosolveOptimizer` has been generalized to arbitrary frequency spectra

>   in the cost function. Also note the changes in behaviour listed under _Breaking
>   changes_.

Reviewer: ```suggestion
  in the cost function. Also note the changes in behaviour listed under *Breaking
  changes*.
```'
Probably makes no difference, but I am always paranoid of the md -> rest conversion!

<h3>Breaking changes</h3>

* The behaviour of `RotosolveOptimizer` has been changed regarding
  its keyword arguments.
  [(#2081)](https://github.com/PennyLaneAI/pennylane/pull/2081)

  The keyword arguments `optimizer` and `optimizer_kwargs` for the
  `RotosolveOptimizer` have been renamed to `substep_optimizer`
  and `substep_kwargs`, respectively. Furthermore they have been
  moved from `step` and `step_and_cost` to the initialization `__init__`.

  The keyword argument `num_freqs` has been renamed to `nums_frequency`
  and is expected to take a different shape now:
  Previously, it was expected to be an `int` or a list of entries, with
  each entry in turn being either an `int` or a `list` of `int` entries.
  Now the expected structure is a nested dictionary, matching the
  formatting expected by
  [qml.fourier.reconstruct](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.fourier.reconstruct.html)
  This also matches the expected formatting of the new keyword arguments
  `spectra` and `shifts`.

  For more details, see the

>   [RotosolveOptimizer documentation](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.RotosolveOptimizer.html).

Reviewer: nice!
